tokenizers/src/tokenizer/mod.rs

```diff
@@ -1131,7 +1131,7 @@ where
                 let pre_tokenized = self.do_pre_tokenize(normalized)?;
                 Ok(pre_tokenized
                     .get_splits(OffsetReferential::Original, OffsetType::Byte)
-                    .into_iter()
+                    .into_maybe_par_iter()
```

Comment:
This is parallelizing the `get_splits` method, not the pretokenization itself. 

Shouldn't you modify `sequences.map` instead ?

Reply:
I think you are right.
If I read the code correctly, it would at best parallelize downstream of the pretokenization and not the pretokenization itself.
I don't think this part is the bottleneck and if it is, #1259 might provide a solution.

Since as used in this part of the code `sequences` is an already iterator(as per function defintion) and there is only map here.

I am looking upstream to where the iterator is fed into .train() and it seems to be here

https://github.com/chris-ha458/tokenizers/blob/21b8cbee3a3b02bb6cce9c129196b8fbdbff0c8e/bindings/python/src/tokenizer.rs#LL1221C42-L1221C60

```
        py.allow_threads(|| {
            ResultShunt::process(buffered_iter, |iter| {
                self.tokenizer
                    .train(&mut trainer, MaybeSizedIterator::new(iter, length))
                    .map(|_| {})
                    .map_err(|e| exceptions::PyException::new_err(e.to_string()))
            })?
        })
```
And the part that constructs that `|iter|` appeares right above.
```
let buffered_iter = PyBufferedIterator::new(
            iterator,
            |element| {
                // Each element of the iterator can either be:
                //  - An iterator, to allow batching
                //  - A string
                if let Ok(s) = element.downcast::<PyString>() {
                    itertools::Either::Right(std::iter::once(s.to_str().map(|s| s.to_owned())))
                } else {
                    match element.iter() {
                        Ok(iter) => itertools::Either::Left(
                            iter.map(|i| i?.extract::<String>())
                                .collect::<Vec<_>>()
                                **.into_iter(), # Change this into into_maybe_par_iter()?**
                        ),
                        Err(e) => itertools::Either::Right(std::iter::once(Err(e))),
                    }
                }
            },
            256,
        )?;
```
My hope is that changing the line with
`**.into_iter(), # Change this into into_maybe_par_iter()?**`
would do the trick.

As far as i can tell there isn't anyplace downstream that uses any sideeffects or shared state so
it would work, and the type system and tests would help guide me if I'm wrong.

Reply:
Okay this is much much more involved than I had hoped.
This is also uncomfortably close to the GIL